components/ai_chat/core/browser/conversation_handler.cc

```diff
@@ -1561,6 +1575,58 @@ void ConversationHandler::SetTemporary(bool temporary) {
   metadata_->temporary = temporary;
 }
 
+void ConversationHandler::MaybeSwitchToVisionModel(
+    const std::optional<std::vector<mojom::UploadedFilePtr>>& uploaded_files) {
+  if (uploaded_files && !uploaded_files->empty() &&
+      std::ranges::any_of(
+          uploaded_files.value(), [](const mojom::UploadedFilePtr& file) {
+            return file->type == mojom::UploadedFileType::kImage ||
+                   file->type == mojom::UploadedFileType::kScreenshot;
+          })) {
+    auto* current_model =
+        model_service_->GetModel(metadata_->model_key.value_or("").empty()
+                                     ? model_service_->GetDefaultModelKey()
+                                     : metadata_->model_key.value());
+    if (!current_model->vision_support) {
+      ChangeModel(ai_chat_service_->IsPremiumStatus()
+                      ? features::kAIModelsPremiumVisionDefaultKey.Get()
+                      : features::kAIModelsVisionDefaultKey.Get());
+    }
+  }
+}
+
+void ConversationHandler::OnScreenshotsTaken(
+    GetPageContentCallback callback,
+    std::string contents_text,
+    bool is_video,
+    std::optional<std::vector<mojom::UploadedFilePtr>> screenshots) {
+  // Screenshots have been taken and are already in UploadedFile format
+  // Update the existing conversation turn with these uploaded files
+
+  if (contents_text.empty() && screenshots && !screenshots->empty()) {
+    // Auto-switch to vision model if needed
+    MaybeSwitchToVisionModel(screenshots);
+
+    // Update the last conversation turn with the uploaded files
+    if (!chat_history_.empty() &&
+        chat_history_.back()->character_type == mojom::CharacterType::HUMAN) {
+      // Append screenshots to existing uploaded files instead of replacing them
+      if (!chat_history_.back()->uploaded_files) {
+        chat_history_.back()->uploaded_files.emplace();
```

Comment:
Took me a while to click that this is calling `emplace` on the optional not the `vector` :laughing: